### ISSUE TITLE
nan loss after a number of steps (epochs)

### ISSUE THREAD

**Issue by tonydifranco** [fferroni/PhasedLSTM-Keras#11]
I am able to train a regression network without error using the base keras `LSTM` layer but always seem to run into a `nan` loss (mse) after what starts out as very promising results using this `PLSTM` layer.  I have tried most of the recommendations in these two keras issues [#2134](https://github.com/fchollet/keras/issues/2134) and [#1244](https://github.com/fchollet/keras/issues/1244) but nothing seems to help.  Do you have any troubleshooting recommendations concerning this issue using your PLSTM implementation?

**edit**
I was able to get through 10x as many steps using the `adamax` optimizer (as opposed to `rmsprop` or `adam`), increased layer sizes all around, and an extra dense layer with a `tanh` activation.  Unfortunately, the loss still went to `nan`.  Again, the training loss was significantly better than anything I was able to squeeze out of the plain LSTM approach. :confused:

**Comment by fferroni** [comment]
Hi,

I did not experience this with PLSTM.

- Have you tried using something like clipvalue=1 in the optimizer?
- Have you used Adam?
- A robust regression loss like Huber loss?

It's hard for me to pin point exactly why it's happening for you with no details.

**Comment by tonydifranco** [comment]
-yes, I did try `clipvalue` and `clipnorm` in the optimizer
-yes, I tried `Adam`, `Adamax`, `RMSprop`,  and `Nadam`
-no, I have not tried `huber` loss because regular LSTM does not result in `nan` loss

I am using `fit_generator` and was able to get through more total steps and epochs before the `nan` loss by decreasing `steps_per_epoch`.  I can provide a few more details.  The general architecture I have been using is close to the following:

```
input_1 = Input(shape=(None, 1297))
input_2 = Input(shape=(178,))
lstm_1 = PLSTM(256)(input_1)
concat_1 = concatenate([lstm_1, input_2])
dense_1 = Dense(256, activation='relu')(concat_1)
dense_2 = Dense(512, activation='relu')(dense_1)
dense_3 = Dense(128, activation='tanh')(dense_2)
output_1 = Dense(1, activation='linear', name='output_1')(dense_3)
model = Model(inputs=[input_2, input_1], outputs=[output_1])
model.compile(loss='mean_squared_error', optimizer='adam')
model.fit_generator(train, steps_per_epoch=1000, epochs=20)
```
The response is a log transformed version of a heavy right-tailed (and non-negative) distribution.   The number of time steps for each training example is variable (hence the `None` in the `Input` layer) and can regularly exceed 300 events.  I have not tried constant time steps with padding yet b/c I believe this structure represents the problem more appropriately.

**Comment by fferroni** [comment]
Sorry I am not sure. The only time I have experienced nan losses, apart from obvious mistakes, is if I do not use a non_neg constraint in my time kernel, which should be the default anyway https://github.com/fferroni/PhasedLSTM-Keras/blob/master/phased_lstm_keras/PhasedLSTM.py#L143
I guess if you are not padding and always feeding 1 sequence at a time, you might want to keep your learning rate quite low. I would still also try using huber loss. Also, a sanity check might be to assert finite inputs in your generator :p

**Comment by tonydifranco** [comment]
thanks... I haven't run into the issue for a while now as I have put a little more effort into the architecture and inputs. great module!

**Comment by ruiatelsevier** [comment]
I have got the same problem. It produces nan after a number of steps.

**Comment by Samyarrahimi** [comment]
I had the nan loss problem in image data, I was using **np.empty** for generating the batches of images and looks like it was the reason of the nan loss, changed **np.empty** to **np.zeros** and problem solved.
